core/src/main/java/org/apache/iceberg/encryption/EnvelopeEncryptionManager.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.encryption;
+
+import java.nio.ByteBuffer;
+import java.security.SecureRandom;
+import java.util.Map;
+import org.apache.iceberg.common.DynConstructors;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+/**
+ * Encryption manager that performs envelope encryption.
+ * <p>
+ */
+public class EnvelopeEncryptionManager implements EncryptionManager {
+  private final EnvelopeConfig dataEncryptionConfig;
+  private final boolean pushdown;
+  private final KmsClient kmsClient;
+  private final int dataKeyLength;
+
+  private transient volatile SecureRandom workerRNG = null;
+
+  public EnvelopeEncryptionManager(
+      boolean pushdown,  EnvelopeConfig dataEncryptionConfig,
+      KmsClient kmsClient, int dataKeyLength) {
+    if (!pushdown) {
+      throw new UnsupportedOperationException("EnvelopeEncryptionManager currently supports only pushdown " +
+          "data encryption / decryption");
+    }
+    this.dataEncryptionConfig = Preconditions.checkNotNull(dataEncryptionConfig,
+        "Cannot create EnvelopeEncryptionManager because data encryption config is not passed");
+    Preconditions.checkNotNull(dataEncryptionConfig.kekId(),
+        "Cannot create EnvelopeEncryptionManager because table key encryption key ID is not specified");
+    this.kmsClient = Preconditions.checkNotNull(kmsClient,
+        "Cannot create EnvelopeEncryptionManager because KmsClient is null");
+    this.pushdown = pushdown;
+    this.dataKeyLength = dataKeyLength;
+  }
+
+  @Override
+  public EncryptedOutputFile encrypt(OutputFile rawOutput) {
+    EnvelopeMetadata metadata = generateEnvelopeMetadata(dataEncryptionConfig);
+
+    // Pushdown to data formats with native encryption support (such as Parquet)
+    NativeFileCryptoParameters nativeEncryptParams = NativeFileCryptoParameters.create(metadata.dek())
```

Comment:
I'd prefer to check the class member `pushdown` even though non-pushdown is not supported at this moment.

Reply:
```
if(pushdown) {
  return nativeEncrypt(output);
} else {
  throw new Exception();
}

Reply:
SGTM